src/ngraph_encapsulate_op.cc

```diff
@@ -360,13 +360,23 @@ class NGraphEncapsulateOp : public OpKernel {
           }
         } else {
           if (last_tv != nullptr) {
+            if (current_src_ptr == last_src_ptr) {
+              if (m_freshness_tracker->IsFresh(current_src_ptr, ng_function)) {
+                last_tv->set_stale(false);
+              } else {
+                last_tv->set_stale(true);
+              }
+            }
             current_tv = last_tv;
           } else {
             current_tv = m_ng_backend->create_tensor(ng_element_type, ng_shape);
+            current_tv->set_stale(true);
+          }
+          if (current_tv->get_stale()) {
+            current_tv->write(
+                current_src_ptr, 0,
+                current_tv->get_element_count() * ng_element_type.size());
           }
-          current_tv->write(
```

Comment:
The if else of "CPU" and "non-CPU" is starting to look very similar.

We can rearrange the if-else tree in lines 339 to 378 with some other structure to avoid repeating. For example:

```
bool last_tv_is_null = last_tv == nullptr;
bool curr_eq_last = current_src_ptr == last_src_ptr;
bool is_cpu = s_ng_backend_name == "CPU";

// set staleness of last tv if needed
if (!last_tv_is_null && curr_eq_last)
  last_tv->set_stale(!m_freshness_tracker->IsFresh(current_src_ptr, ng_function));

// last can be assigned to curr if last is not null
// except in CPU, which needs an additional check: curr_eq_last
if (!last_tv_is_null && ((is_cpu && curr_eq_last) || (!is_cpu))){
  current_tv = last_tv;
} else {
  // last could not be used, create a new tensor
  if (is_cpu)
    current_tv = m_ng_backend->create_tensor(ng_element_type, ng_shape, current_src_ptr);
  else
    current_tv = m_ng_backend->create_tensor(ng_element_type, ng_shape);
  current_tv->set_stale(true);
}

// write to current_tv if it is stale for non-CPU
if (!is_cpu && current_tv->get_stale())
  current_tv->write(current_src_ptr, 0, current_tv->get_element_count() * ng_element_type.size());

```

If nothing, we should replace the 2 
```
if (m_freshness_tracker->IsFresh(current_src_ptr, ng_function)) {
  last_tv->set_stale(false);
} else {
  last_tv->set_stale(true);
}
```
with
```
last_tv->set_stale(!m_freshness_tracker->IsFresh(current_src_ptr, ng_function));
```